include/deal.II/matrix_free/tools.h

```diff
@@ -157,6 +157,183 @@ namespace MatrixFreeTools
     const unsigned int first_selected_component = 0);
 
 
+
+  /**
+   * A wrapper around MatrixFree to help users to deal with DoFHandlers,
+   * where there are cells without degrees of freedom, i.e., with
+   * FENothing assigned to them. In the following we call such cells
+   * invalid. All other cells are valid. In contrast to MatrixFree,
+   * this class skips invalid cells and faces between valid and
+   * invalid cells are treated as boundary faces.
+   *
+   * The class helps to implement "element birth and death technique",
+   * in which context valid cells are alive and invalid cells are dead.
+   */
+  template <int dim,
+            typename Number,
+            typename VectorizedArrayType = VectorizedArray<Number>>
+  class ElementBirthAndDeathMatrixFree
+  {
+  public:
+    /**
+     * Struct that helps to configure ElementBirthAndDeathMatrixFree.
+     */
+    struct AdditionalData
+    {
+      /**
+       * Constructor.
+       */
+      AdditionalData(const unsigned int dof_index = 0)
+        : dof_index(dof_index)
+      {}
+
+      /**
+       * Index of the DoFHandler within MatrixFree to be used.
+       */
+      unsigned int dof_index;
+    };
+
+    /**
+     * Reinitialize class based on a given MatrixFree instance.
+     * Particularly, the index of the valid FE is determined.
+     */
+    void
+    reinit(const MatrixFree<dim, Number, VectorizedArrayType> &matrix_free,
+           const AdditionalData &additional_data = AdditionalData())
+    {
+      this->matrix_free = &matrix_free;
+
+      std::set<unsigned int> valid_fe_indices;
+
+      const auto &fe_collection =
+        matrix_free.get_dof_handler(additional_data.dof_index)
+          .get_fe_collection();
+
+      for (unsigned int i = 0; i < fe_collection.size(); ++i)
+        if (fe_collection[i].n_dofs_per_cell() > 0)
+          valid_fe_indices.insert(i);
+
+      AssertDimension(valid_fe_indices.size(), 1);
```

Comment:
Can we allow more than one `valid_fe_indices`? Or create an `invalid_fe_index` and check for it later?

Reply:
> Also, I guess what you need in your case is to not loop over faces neighboring FE_Nothing.

Yes, this should only be optional from my perspective. For me this class can definitely be useful, currently I skip FENothing cells (or cell batches) manually.